File: DifferLib/Highlighter.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace DifferLib
{
    public static class Highlighter
    {
        public static List<(HighlightedLine Before, HighlightedLine After)> Highlight(string before, string after, List<SubstringDescriptor> deleteDescs, List<SubstringDescriptor> insertDescs)
        {
            if (before == null) throw new ArgumentNullException(nameof(before));
            if (after == null) throw new ArgumentNullException(nameof(after));
            if (deleteDescs == null) throw new ArgumentNullException(nameof(deleteDescs));
            if (insertDescs == null) throw new ArgumentNullException(nameof(insertDescs));

            var highlightedBefore = ComputeLines(before, deleteDescs);
            var highlightedAfter = ComputeLines(after, insertDescs);

            var pairs = LinesMatcher(highlightedBefore, highlightedAfter);
            return pairs;
        }

        private static List<(HighlightedLine Before, HighlightedLine After)> LinesMatcher(HighlightedLine[] before, HighlightedLine[] after)
        {
            var pairs = new List<(HighlightedLine Before, HighlightedLine After)>();

            var beforeIndex = 0;
            var afterIndex = 0;
            var beforeRetainedLineStart = 0;
            var afterRetainedLineStart = 0;

            while (beforeIndex < before.Length || afterIndex < after.Length)
            {
                while (IsFullyHighlighted(before, beforeIndex))
                {
                    var shouldGoInPair = false;
                    if (IsFullyHighlighted(after, afterIndex))
                    {
                        shouldGoInPair = true;
                        afterIndex += 1;
                    }

                    pairs.Add((before[beforeIndex++], shouldGoInPair ? after[afterIndex - 1] : null));
                }

                while (IsFullyHighlighted(after, afterIndex))
                {
                    pairs.Add((null, after[afterIndex]));
                    afterIndex += 1;
                }

                if (beforeRetainedLineStart != afterRetainedLineStart)
                {
                    throw new InvalidProgramException($"After can't be different from before - {afterRetainedLineStart} against {beforeRetainedLineStart}");
                }

                if (beforeIndex < before.Length)
                {
                    var retained = before[beforeIndex].Regular.Sum(b => b.Length);
                    beforeRetainedLineStart += retained;

                    if (afterIndex >= after.Length)
                    {
                        throw new InvalidProgramException($"There are not enough after lines. {beforeIndex}:{before.Length} - {afterIndex}:{after.Length}");
                    }

                    afterRetainedLineStart += retained;
                    pairs.Add((before[beforeIndex++], after[afterIndex++]));
                }
                else
                {
                    if (afterIndex < after.Length)
                    {
                        throw new InvalidProgramException($"There are not enough before lines. {beforeIndex}:{before.Length} - {afterIndex}:{after.Length}");
                    }
                }
            }

            return pairs;
        }

        private static bool IsFullyHighlighted(HighlightedLine[] lines, int index)
        {
            return index < lines.Length && lines[index].Regular.Sum(b => b.Length) == 0;
        }

        private static HighlightedLine[] ComputeLines(
            string source,
            List<SubstringDescriptor> descs)
        {
            var sourceIndex = 0;
            var curDescIndex = 0;
            var lines = SplitByLines(source);

            var highlightedLines = new HighlightedLine[lines.Count];

            for (int i = 0; i < lines.Count; ++i)
            {
                var line = lines[i];

                var highlightedLineFactory = new HighlightedLineFactory();

                if (curDescIndex < descs.Count && descs[curDescIndex].Start < line.End)
                {
                    do
                    {
                        var isLeftover = descs[curDescIndex].Start < line.Start;
                        if (!isLeftover)
                        {
                            highlightedLineFactory.Add(source, ref sourceIndex, descs[curDescIndex].Start - sourceIndex);
                        }

                        if (descs[curDescIndex].End <= line.End)
                        {
                            highlightedLineFactory.AddHighlighted(source, ref sourceIndex, descs[curDescIndex].End - sourceIndex);
                            curDescIndex += 1;
                        }
                        else
                        {
                            highlightedLineFactory.AddHighlighted(source, ref sourceIndex, line.End - sourceIndex);
                            break;
                        }
                    }
                    while (curDescIndex < descs.Count && descs[curDescIndex].Start < line.End);
                }

                if (line.End - sourceIndex != 0)
                {
                    highlightedLineFactory.Add(source, ref sourceIndex, line.End - sourceIndex);
                }

                highlightedLines[i] = highlightedLineFactory.Create();
            }

            return highlightedLines;
        }

        private static IReadOnlyList<Line> SplitByLines(string input)
        {
            var ret = new List<Line>();
            for (int i = 0; i < input.Length;)
            {
                var nextLineStartIndex = input.IndexOf('\n', i) + 1;
                if (nextLineStartIndex == 0) nextLineStartIndex = input.Length;
                ret.Add(new Line(i, nextLineStartIndex - i));
                i = nextLineStartIndex;
            }
            return ret;
        }

        private struct Line
        {
            public int Start { get; }
            public int End => Start + Length;
            public int Length { get; }
            public Line(int start, int length)
            {
                Start = start;
                Length = length;
            }
        }
    }
}

File: WebVisualizer/wwwroot/DifferLibJS/Highlighter.js
class HighlightedLine
{
    get blocks() { return this._blocks; }
    get regular() { return this._blocks.filter((_, index) => index % 2 == 0); }
    get highlighted() { return this._blocks.filter((_, index) => index % 2 == 1); }

    get hasHighlighted() { return this._blocks.length > 1; }

    constructor(blocks)
    {
        if (blocks == null) throw "";
        this._blocks = blocks;
    }
}

class HighlightedLineFactory
{
    constructor()
    {
        this._blocks = new Array();
    }

    create() { return new HighlightedLine(this._blocks); }

    addHighlighted(source, start, length)
    {
        if (this._blocks.length == 0)
        {
            this._blocks.push("");
        }

        if (this._blocks.length % 2 == 0)
        {
            throw `Can't insert highlighted block. Blocks types must alternate. Current number of blocks is ${_blocks.length}.`;
        }

        this._blocks.push(source.substring(start, start + length));
        return start + length;
    }

    add(source, start, length)
    {
        if (this._blocks.length % 2 == 1)
        {
            throw `Can't insert retained block. Blocks types must alternate. Current number of blocks is ${_blocks.length}.`;
        }

        this._blocks.push(source.substring(start, start + length));
        return start + length;
    }
}

class Line
{
    get start() { return this._start; }
    get end() { return this._start + this._length; }
    get length() { return this._length; }

    constructor(start, length)
    {
        this._start = start;
        this._length = length;
    }
}

class Highlighter
{
    static highlight(before, after, deleteDescs, insertDescs)
    {
        if (before == null) throw "";
        if (after == null) throw "";
        if (deleteDescs == null) throw "";
        if (insertDescs == null) throw "";

        const highlightedBefore = Highlighter.computeLines(before, deleteDescs);
        const highlightedAfter = Highlighter.computeLines(after, insertDescs);

        const pairs = Highlighter.linesMatcher(highlightedBefore, highlightedAfter);
        return pairs;
    }

    static linesMatcher(before, after)
    {
        let pairs = new Array();

        let beforeIndex = 0;
        let afterIndex = 0;
        let beforeRetainedLineStart = 0;
        let afterRetainedLineStart = 0;

        while (beforeIndex < before.length || afterIndex < after.length)
        {
            while (Highlighter.isFullyHighlighted(before, beforeIndex))
            {
                let shouldGoInPair = false;
                if (Highlighter.isFullyHighlighted(after, afterIndex))
                {
                    shouldGoInPair = true;
                    afterIndex += 1;
                }

                pairs.push({
                    before: before[beforeIndex++],
                    after: shouldGoInPair ? after[afterIndex - 1] : null
                });
            }

            while (Highlighter.isFullyHighlighted(after, afterIndex))
            {
                pairs.push({
                    before: null,
                    after: after[afterIndex]
                });
                afterIndex += 1;
            }

            if (beforeRetainedLineStart != afterRetainedLineStart)
            {
                throw `After can't be different from before - ${afterRetainedLineStart} against ${beforeRetainedLineStart}`;
            }

            if (beforeIndex < before.length)
            {
                const retained = before[beforeIndex].regular.reduce((accum, block) => accum + block.length, 0);
                beforeRetainedLineStart += retained;

                if (afterIndex >= after.length)
                {
                    throw `There are not enough after lines. ${beforeIndex}:${before.length} - ${afterIndex}:${after.length}`;
                }

                afterRetainedLineStart += retained;
                pairs.push({
                    before: before[beforeIndex++],
                    after: after[afterIndex++]
                });
            }
            else
            {
                if (afterIndex < after.length)
                {
                    throw `There are not enough before lines. ${beforeIndex}:${before.length} - ${afterIndex}:${after.length}`;
                }
            }
        }

        return pairs;
    }

    static isFullyHighlighted(lines, index)
    {
        if (index < lines.length) {
            const regularLength = lines[index].regular.reduce((accum, block) => accum + block.length, 0);
            return regularLength == 0;
        }
        return false;
    }

    static computeLines(source, descs)
    {
        let sourceIndex = 0;
        let curDescIndex = 0;
        const lines = Highlighter.splitByLines(source);

        let highlightedLines = new Array(lines.length);

        for (let i = 0; i < lines.length; ++i)
        {
            const line = lines[i];

            const highlightedLineFactory = new HighlightedLineFactory();

            if (curDescIndex < descs.length && descs[curDescIndex].start < line.end)
            {
                do
                {
                    var isLeftover = descs[curDescIndex].start < line.start;
                    if (!isLeftover)
                    {
                        sourceIndex = highlightedLineFactory.add(source, sourceIndex, descs[curDescIndex].start - sourceIndex);
                    }

                    if (descs[curDescIndex].end <= line.end)
                    {
                        sourceIndex = highlightedLineFactory.addHighlighted(source, sourceIndex, descs[curDescIndex].end - sourceIndex);
                        curDescIndex += 1;
                    }
                    else
                    {
                        sourceIndex = highlightedLineFactory.addHighlighted(source, sourceIndex, line.end - sourceIndex);
                        break;
                    }
                }
                while (curDescIndex < descs.length && descs[curDescIndex].start < line.end);
            }

            if (line.end - sourceIndex != 0)
            {
                sourceIndex = highlightedLineFactory.add(source, sourceIndex, line.end - sourceIndex);
            }

            highlightedLines[i] = highlightedLineFactory.create();
        }

        return highlightedLines;
    }

    static splitByLines(input)
    {
        const ret = new Array();
        for (let i = 0; i < input.length;)
        {
            let nextLineStartIndex = input.indexOf('\n', i) + 1;
            if (nextLineStartIndex == 0) nextLineStartIndex = input.length;
            ret.push(new Line(i, nextLineStartIndex - i));
            i = nextLineStartIndex;
        }
        return ret;
    }
}
